lib/discourse_redis.rb

```diff
@@ -4,139 +4,251 @@
+      unless info.include?(MASTER_LOADED_STATUS)
+        warn "Master not alive, status is loading"
+        return false
+      end
 
-            self.master = true
-            slave_client = ::Redis::Client.new(@slave_config)
+      unless info.include?(MASTER_ROLE_STATUS)
+        warn "Master not alive, role != master"
+        return false
+      end
 
-            CONNECTION_TYPES.each do |connection_type|
-              slave_client.call([:client, [:kill, 'type', connection_type]])
-            end
+      true
+    end
 
-            MessageBus.keepalive_interval = @message_bus_keepalive_interval
-            Discourse.clear_readonly!
-            Discourse.request_refresh!
-            success = true
-          ensure
-            slave_client&.disconnect
-          end
+    def fallback
+      warn "Killing connections to slave..."
+
+      slave_client = connect(@slave_config)
+
+      begin
+        CONNECTION_TYPES.each do |connection_type|
+          slave_client.call([:client, [:kill, 'type', connection_type]])
         end
-      rescue => e
-        logger.warn "#{log_prefix}: Connection to Master server failed with '#{e.message}'"
+      rescue Redis::ConnectionError, Redis::CannotConnectError, RuntimeError => ex
+        raise ex if ex.class == RuntimeError && ex.message != "Name or service not known"
+        warn "Attempted a redis fallback, but connection to slave failed"
       ensure
-        master_client&.disconnect
+        slave_client.disconnect
+      end
+    end
+
+    private
+
+    def connect(config)
+      config = config.dup
+      config.delete(:connector)
+      ::Redis::Client.new(config)
+    end
+
+    def log_prefix
+      @log_prefix ||= begin
+        master_string = "#{@master_config[:host]}:#{@master_config[:port]}"
+        slave_string = "#{@slave_config[:host]}:#{@slave_config[:port]}"
+        "RedisStatus master=#{master_string} slave=#{slave_string}"
       end
+    end
+
+    def warn(message)
+      Rails.logger.warn "#{log_prefix}: #{message}"
+    end
+  end
 
-      success
+  class FallbackHandler
+    def initialize(log_prefix, redis_status, execution)
+      @log_prefix = log_prefix
+      @redis_status = redis_status
+      @mutex = execution.new_mutex
+      @execution = execution
+      @master = true
+      @event_handlers = []
     end
 
-    def master
-      synchronize { @master }
+    def add_callbacks(handler)
+      @mutex.synchronize do
+        @event_handlers << handler
+      end
     end
 
-    def master=(args)
-      synchronize do
-        @master = args
+    def use_master?
+      master = @mutex.synchronize { @master }
+      if !master
+        return false
+      elsif safe_master_alive?
+        return true
+      else
+        was_reset =
+          @mutex.synchronize do
+            if @master
+              @master = false
+              trigger(:down)
+              true
+            else
+              false
+            end
+          end
```

Comment:
I would extract that code to separate method something like
```
def start_reset
   @mutex.synchronize do
            if @master
              @master = false
              trigger(:down)
              true
            else
              false
            end
          end
end
```
and then you would be able to change later code to be like
```
        if start_reset
          @execution.spawn do
            loop do
```